File: BaekjoonAlgorithm/BA_14890/BJAlgo_14890.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BJAlgo_14890 {
    private static int n;
    private static int l;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());

        int[][] map = new int[n][n];
        for (int i = 0; i < n; i++) {
            map[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }

        int path = 0;

        for (int i = 0; i < n; i++) {
            int[] tmp = map[i];
            if (findPath(tmp)) {
                path += 1;
            }

            tmp = new int[n];
            for (int j = 0; j < n; j++) {
                tmp[j] = map[j][i];
            }

            if (findPath(tmp)) {
                path += 1;
            }
        }

        System.out.println(path);
    }

    private static boolean findPath(int[] map) {
        boolean answer = true;
        int last = map[0];
        int count = 1;

        int cnt = 1;
        while (cnt < n) {
            if (last == map[cnt]) {
                count += 1;
                cnt += 1;
            } else if (map[cnt] - last == 1) {
                if (count < l) {
                    answer = false;
                    break;
                }

                last = map[cnt];
                count = 1;
                cnt += 1;
            } else if (last - map[cnt] == 1) {
                if (n - cnt >= l) {
                    for (int i = cnt + 1; i < cnt + l; i++) {
                        if (map[cnt] != map[i]) {
                            answer = false;
                            break;
                        }
                    }

                    if (!answer) {
                        break;
                    }

                    cnt += l;
                    last = map[cnt - 1];
                    count = 0;
                } else {
                    answer = false;
                    break;
                }
            } else {
                answer = false;
                break;
            }
        }

        return answer;
    }

}

File: BaekjoonAlgorithm/202102/BA_14890/BJAlgo_14890.py
from sys import stdin


def find_path(p):
    answer = True
    last = p[0]
    count = 1

    cnt = 1
    while cnt < n:
        if last == p[cnt]:
            count += 1
            cnt += 1
        elif p[cnt] - last == 1:
            if count < l:
                answer = False
                break

            last = p[cnt]
            count = 1
            cnt += 1
            continue
        elif last - p[cnt] == 1:
            if n - cnt >= l:
                for a in range(cnt + 1, cnt + l):
                    if p[cnt] != p[a]:
                        answer = False
                        break
                if not answer:
                    break

                cnt = cnt + l
                last = p[cnt - 1]
                count = 0
            else:
                answer = False
                break
        else:
            answer = False
            break

    return answer


n, l = map(int, stdin.readline().split())
matrix = []
for _ in range(n):
    matrix.append(list(map(int, stdin.readline().split())))

path = 0

for i in range(n):
    tmp = matrix[i]
    if find_path(tmp):
        path += 1

    tmp = []
    for j in range(n):
        tmp.append(matrix[j][i])

    if find_path(tmp):
        path += 1

print(path)
